File: iq/18/cycle/is_test.cc
// Copyright 2026 Samvel Khalatyan. All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

//
// clang-format off-next-line
//go:build ignore

#include "iq/18/cycle/is.h"

#include <memory>

#include "gtest/gtest.h"

namespace iq::cycle {
namespace {

TEST(IsTest, empty) {
  std::shared_ptr<Node> list;

  const bool got = Is(list);

  EXPECT_FALSE(got);
}

TEST(IsTest, oneItemNoCycle) {
  std::shared_ptr<Node> list = std::make_shared<Node>(1);

  const bool got = Is(list);

  EXPECT_FALSE(got);
}

TEST(IsTest, oneItemCycle) {
  std::shared_ptr<Node> list = std::make_shared<Node>(1);
  list->next = list;

  const bool got = Is(list);

  EXPECT_TRUE(got);
}

TEST(IsTest, twoItemsNoCycle) {
  std::shared_ptr<Node> list =
      std::make_shared<Node>(1, std::make_shared<Node>(2));

  const bool got = Is(list);

  EXPECT_FALSE(got);
}

TEST(IsTest, twoItemsCycleToFirst) {
  std::shared_ptr<Node> list =
      std::make_shared<Node>(1, std::make_shared<Node>(2));
  list->next->next = list;

  const bool got = Is(list);

  EXPECT_TRUE(got);
}

TEST(IsTest, twoItemsCycleToSecond) {
  std::shared_ptr<Node> list =
      std::make_shared<Node>(1, std::make_shared<Node>(2));
  list->next->next = list->next;

  const bool got = Is(list);

  EXPECT_TRUE(got);
}

}  // namespace
}  // namespace iq::cycle

File: iq/18/cycle/is_test.go
// Copyright 2026 Samvel Khalatyan. All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package cycle_test

import (
	"fmt"
	"testing"

	"github.com/skhal/lab/iq/18/cycle"
)

type testNode cycle.Node

func (node *testNode) String() string {
	var nn []int
	visits := make(map[*testNode]struct{})
	for ; node != nil; node = (*testNode)(node.Next) {
		nn = append(nn, node.Val)
		if _, ok := visits[node]; ok {
			// cycle
			break
		}
		visits[node] = struct{}{}
	}
	return fmt.Sprint(nn)
}

func TestIs_empty(t *testing.T) {
	var node *cycle.Node

	got := cycle.Is(node)

	if want := false; got != want {
		t.Errorf("cycle.Is(%s) = %v; want %v", (*testNode)(node), got, want)
	}
}

func TestIs_oneItemNoCycle(t *testing.T) {
	node := &cycle.Node{Val: 1}

	got := cycle.Is(node)

	if want := false; got != want {
		t.Errorf("cycle.Is(%s) = %v; want %v", (*testNode)(node), got, want)
	}
}

func TestIs_oneItemCycle(t *testing.T) {
	node := &cycle.Node{Val: 1}
	node.Next = node

	got := cycle.Is(node)

	if want := true; got != want {
		t.Errorf("cycle.Is(%s) = %v; want %v", (*testNode)(node), got, want)
	}
}

func TestIs_twoItemsNoCycle(t *testing.T) {
	node := &cycle.Node{
		Val: 1,
		Next: &cycle.Node{
			Val: 2,
		},
	}

	got := cycle.Is(node)

	if want := false; got != want {
		t.Errorf("cycle.Is(%s) = %v; want %v", (*testNode)(node), got, want)
	}
}

func TestIs_twoItemsCycleToFirst(t *testing.T) {
	node := &cycle.Node{
		Val: 1,
		Next: &cycle.Node{
			Val: 2,
		},
	}
	node.Next.Next = node

	got := cycle.Is(node)

	if want := true; got != want {
		t.Errorf("cycle.Is(%s) = %v; want %v", (*testNode)(node), got, want)
	}
}

func TestIs_twoItemsCycleToSecond(t *testing.T) {
	node := &cycle.Node{
		Val: 1,
		Next: &cycle.Node{
			Val: 2,
		},
	}
	node.Next.Next = node.Next

	got := cycle.Is(node)

	if want := true; got != want {
		t.Errorf("cycle.Is(%s) = %v; want %v", (*testNode)(node), got, want)
	}
}
